### ISSUE TITLE
Is this RNN implementation different from the vanilla RNN?

### ISSUE THREAD

**Issue by KinWaiCheuk** [pytorch/tutorials#566]
https://github.com/pytorch/tutorials/blob/master/intermediate_source/char_rnn_classification_tutorial.py

Standard Interpretation
-------------------------

In the original RNN, the hidden state and output are calculated as

[![enter image description here][1]][1]

in other words, we obtain the the output from the hidden state.

According to [Wiki][2], the RNN architecture can be unfolded like this
[![vani][3]][3]

And the code I have been using is like:

    class Model(nn.Module):
        def __init__(self, input_size, output_size, hidden_dim, n_layers):
            super(Model, self).__init__()
            self.hidden_dim = hidden_dim
            self.rnn = nn.RNN(input_size, hidden_dim, 1)   
            self.fc = nn.Linear(hidden_dim, output_size)
        
        def forward(self, x):
            batch_size = x.size(0)
    
            out, hidden = self.rnn(x)
            
            # getting output from the hidden state
            out = out..view(-1, self.hidden_dim)
            out = self.fc(out)
            
            return out, hidden

RNN as "pure" feed-forward layers
-------------------------
But in this tutorial the hidden layer calculation is same as the standard interpretation, but the output is is calculated independently from the current hidden state `h`.

To me, the math behind this implementation is:

[![enter image description here][6]][6]

So, this implementation is different from the original RNN implementation?

  [1]: https://i.stack.imgur.com/1IdH7.png
  [2]: https://en.wikipedia.org/wiki/Recurrent_neural_network
  [3]: https://i.stack.imgur.com/aJL7l.png
  [4]: https://pytorch.org/tutorials/intermediate/char_rnn_classification_tutorial.html#creating-the-network
  [5]: https://i.stack.imgur.com/mfjcp.png
  [6]: https://i.stack.imgur.com/M3jgf.gif

@chsasank @zou3519

**Comment by superuser992002** [comment]
i think this implementation is equivalent to the RNN model using ReLU activation on the state. Meaning that ht = Whxt + Uhht-1 + bh and yt = softmax(Wyht+by) now is softmax(Wy(Whxt + Uhht-1)+by) = softmax(WyWhxt + WyUhht-1 + by) if we consider WyWh as new Wy and WyUh as new Uy then the implementation make sense.
